src/ancestry_proof.rs

```diff
@@ -136,109 +135,78 @@ impl<T: Clone + PartialEq, M: Merge<Item = T>> NodeMerkleProof<T, M> {
-    debug_assert!(!nodes.is_empty(), "can't be empty");
-    // (position, hash, height)
-
-    let mut queue: VecDeque<_> = nodes
+    let mut queue: VecDeque<_> = VecDeque::new();
+    for value in nodes
         .into_iter()
         .map(|(pos, item)| (pos, item, pos_height_in_tree(pos)))
-        .collect();
-
-    let mut sibs_processed_from_back = Vec::new();
+    {
+        if !queue.insert_sorted_by(value, cmp_nodes) {
+            return Err(Error::CorruptedProof);
+        }
+    }
 
-    // calculate tree root from each items
     while let Some((pos, item, height)) = queue.pop_front() {
         if pos == peak_pos {
             if queue.is_empty() {
                 // return root once queue is consumed
                 return Ok(item);
-            }
-            if queue
-                .iter()
-                .any(|entry| entry.0 == peak_pos && entry.1 != item)
-            {
+            } else {
                 return Err(Error::CorruptedProof);
             }
-            if queue
-                .iter()
-                .all(|entry| entry.0 == peak_pos && &entry.1 == &item && entry.2 == height)
-            {
-                // return root if remaining queue consists only of duplicate root entries
-                return Ok(item);
-            }
-            // if queue not empty, push peak back to the end
-            queue.push_back((pos, item, height));
-            continue;
         }
         // calculate sibling
         let next_height = pos_height_in_tree(pos + 1);
         let (parent_pos, parent_item) = {
             let sibling_offset = sibling_offset(height);
             if next_height > height {
                 // implies pos is right sibling
-                let (sib_pos, parent_pos) = (pos - sibling_offset, pos + 1);
+                let sib_pos = pos - sibling_offset;
+                let parent_pos = pos + 1;
                 let parent_item = if Some(&sib_pos) == queue.front().map(|(pos, _, _)| pos) {
                     let sibling_item = queue.pop_front().map(|(_, item, _)| item).unwrap();
                     M::merge(&sibling_item, &item)?
-                } else if Some(&sib_pos) == queue.back().map(|(pos, _, _)| pos) {
-                    let sibling_item = queue.pop_back().map(|(_, item, _)| item).unwrap();
-                    M::merge(&sibling_item, &item)?
-                }
-                // handle special if next queue item is descendant of sibling
-                else if let Some(&(front_pos, ..)) = queue.front() {
-                    if height > 0 && is_descendant_pos(sib_pos, front_pos) {
-                        queue.push_back((pos, item, height));
-                        continue;
-                    } else {
-                        return Err(Error::CorruptedProof);
-                    }
                 } else {
+                    // let sibling_item = &proof_iter.next().ok_or(Error::CorruptedProof)?.1;
+                    // M::merge(sibling_item, &item)?
```

Comment:
remove leftover?
```suggestion
```

Reply:
This is the code used in `mmr.rs`. I just thought maybe it's good to keep it commented, just to see the difference between `mmr.rs` and `ancestry_proof.rs` very clear.

Reply:
Done